accepted/capability-checks/capability-checks.md

```diff
@@ -0,0 +1,234 @@
+{
+    [CapabilityCheck(typeof(Registry), nameof(Registry.IsSupported))]
+    [AlwaysAvailabeOn(OSPlatform.Windows)]
+    [AlwaysAvailabeOn(OSPlatform.Linux)]
+    [AlwaysAvailabeOn(OSPlatform.FreeBSD)]
+    [AlwaysUnsupportedOn(WebAssembly)]
+    public sealed class RequiresRegistryCapabilityAttribute : CapabilityAttribute
+    {
+    }
+
+    [RequiresRegistryCapability]
+    public partial class Registry
+    {
+        public static bool IsSupported { get; }
+
+        public static readonly RegistryKey CurrentUser;
+    }
+
+    [RequiresRegistryCapability]
+    public partial class RegistryKey : IDisposable
+    {
+        public RegistryKey OpenSubKey(string path) => throw null;
+        public object GetValue(string name) => throw null;
+        public void Dispose() => throw null;
+    }
+}
+```
+
+### Example: ACLs
+
+```C#
+void Main()
+{
+    if (File.Exits(path))
+    {
+        using (var x = File.OpenRead(path))
+        {
+            if (ObjectSecurity.IsSupported)
+            {
+                var acl = x.GetAccessControl();
+                acl.GiveMePermissions();
+            }
+        }
+    }
+}
+```
+
+```C#
+namespace System.Security.AccessControl
+{
+    [CapabilityCheck(typeof(ObjectSecurity), nameof(ObjectSecurity.IsSupported))]
+    public sealed class RequiresAccessControlCapabilityAttribute : CapabilityAttribute
+    {
+    }
+
+    [RequiresAccessControlCapability]
+    public partial class ObjectSecurity
+    {
+        public static bool IsSupported { get; }
+    }
+
+    [RequiresAccessControlCapability]
+    public partial class FileSecurity
+    {
+    }
+
+    [RequiresAccessControlCapability]
+    public partial class MutexSecurity
+    {
+    }
+}
+
+namespace System.IO
+{
+    public partial class File
+    {
+        [RequiresAccessControlCapability]
+        public static FileSecurity GetAccessControl(string path);
+    }
+
+    public partial class FileInfo
+    {
+        [RequiresAccessControlCapability]
+        public FileSecurity GetAccessControl();
+    }
+}
+
+namespace System.Threading
+{
+    public partial class Mutex
+    {
+        [RequiresAccessControlCapability]
+        public Mutex(bool initiallyOwned, string name, out bool createdNew, MutexSecurity mutexSecurity) => throw null;
+    }
+}
+```
+
+### Example: JIT
+
+```C#
+namespace System.Reflection.Emit
+{
+    [CapabilityCheck(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeSupported))]
+    [CapabilityCheck(typeof(RuntimeFeature), nameof(RuntimeFeature.IsDynamicCodeCompiled))]
+    public partial class RequiresDynamicCodeCapability : CapabilityAttribute
+    {
+    }
+
+    [RequiresDynamicCodeCapability]
+    public partial class DynamicMethod : MethodInfo
+    {
+    }
+
+    [RequiresDynamicCodeCapability]
+    public partial class AssemblyBuilder : Assembly
+    {
+    }
+
+    // Would also mark
+    // - Other *Builder types
+    // - AppDomain.DefineDynamicAssembly
+    // - ILGenerator
+}
+```
+
+## Q & A
```

Comment:
Should we possibly fold the proposed "linker unfriendly" attribute into this as well?: https://github.com/dotnet/runtime/issues/33862
It looks like a good fit.